Libraries/LibWeb/HTML/HTMLFormElement.cpp

```diff
@@ -138,11 +138,19 @@ WebIDL::ExceptionOr<void> HTMLFormElement::submit_form(GC::Ref<HTMLElement> subm
             }
         }
 
-        // FIXME: 4. If the submitter element's no-validate state is false, then interactively validate the constraints
-        //           of form and examine the result. If the result is negative (i.e., the constraint validation concluded
-        //           that there were invalid fields and probably informed the user of this), then:
-        //           1. Set form's firing submission events to false.
-        //           2. Return.
+        // 4. If the submitter element's no-validate state is false, then interactively validate the constraints
+        //    of form and examine the result. If the result is negative (i.e., the constraint validation concluded
+        //    that there were invalid fields and probably informed the user of this), then:
+        //      1. Set form's firing submission events to false.
+        //      2. Return.
+        if (!as<FormAssociatedElement>(*submitter).novalidate_state()) {
```

Comment:
`as<T>()` has the unfortunate side effect that it will crash if the given argument can't be cast to the output type. I found this was causing a crash in: http://wpt.live/html/semantics/forms/form-submission-0/form-double-submit-requestsubmit.html. In this case `submitter` was a `HTMLFormElement`, which isn't a `FormAssociatedElement`.

`as_if<T>()` might be useful here. It returns `null` if casting to the output type fails. Alternatively, `novalidate_state()` could be a moved to `HTMLElement`. I'm not immediately sure what the best approach here is.

Reply:
Ah, nice I had missed that `as_if` function, the test you linked doesn't crash with that 🙂